File: js/controllers/authController.js
/**
 * @module controllers/authController
 * Responsible for defining the server's response to request and responses
 * related to authentication (logging in/out)
 */
const bcrypt = require('bcryptjs');
const jwtService = require('../services/jwtService');
const User = require('../models/userModel');

/**
 * Authenticates a user and generates a JWT token.
 *
 * @param {Object} req - Express request object.
 * @param {Object} res - Express response object.
 */
exports.loginUser = async (req, res) => {
  const { username, password } = req.body;

  try {
    const user = await User.getByUsername(username);

    if (!user) {
      return res.status(400).send('Invalid username or password');
    }

    if (user.isApproved === 0) {
      return res.status(403).send('Your account is awaiting approval.');
    }

    // Verify the entered password
    const validPassword = await bcrypt.compare(password, user.password);
    if (!validPassword) {
      return res.status(400).send('Invalid username or password');
    }

    const token = jwtService.generateToken({ username: user.username, isAdmin: user.isAdmin });
    res.json({ token });
  } catch (error) {
    console.error('Error logging in:', error);
    res.status(500).json({ error: 'Server error' });
  }
};

/** Note: For JWT, logout can be handled on the client-side by deleting the token.  
 * Alternatively, implement token blacklisting on the server-side if necessary.
 */
exports.logoutUser = (req, res) => {
  res.send('Logout successful');
};

// TODO: Password reset logic

File: python/controllers/auth_controller.py
"""
Authentication Controller Module

Responsible for defining the server's response to requests and responses
related to authentication (logging in/out)
"""
import bcrypt
from flask import Blueprint, request, jsonify
from services.jwt_service import JWTService
from models.user_model import UserModel

class AuthController:
    """
    Controller for handling authentication-related operations.
    Uses a class-based approach for better organization and OOP principles.
    """
    
    def __init__(self, user_model=None, jwt_service=None):
        """
        Initialize with dependency injection for better testability.
        
        Args:
            user_model: The user model to use (defaults to UserModel if None)
            jwt_service: The JWT service to use (defaults to JWTService if None)
        """
        self.user_model = user_model or UserModel()
        self.jwt_service = jwt_service or JWTService()
    
    def login_user(self):
        """
        Authenticates a user and generates a JWT token.
        
        Returns:
            JSON response with token or error message
        """
        data = request.get_json()
        username = data.get('username')
        password = data.get('password')
        
        try:
            user = self.user_model.get_by_username(username)
            
            if not user:
                return jsonify({'error': 'Invalid username or password'}), 400
                
            if user.is_approved == 0:
                return jsonify({'error': 'Your account is awaiting approval.'}), 403
                
            # Verify the entered password
            # We need to encode both the input password and stored hash for bcrypt
            valid_password = bcrypt.checkpw(
                password.encode('utf-8'), 
                user.password.encode('utf-8')
            )
            
            if not valid_password:
                return jsonify({'error': 'Invalid username or password'}), 400
                
            token = self.jwt_service.generate_token({
                'username': user.username,
                'is_admin': user.is_admin
            })
            
            return jsonify({'token': token})
            
        except Exception as err:
            print(f"Error logging in: {err}")
            return jsonify({'error': 'Server error'}), 500
            
    def logout_user(self):
        """
        Handles user logout.
        
        Note: For JWT, logout is typically handled client-side by removing the token.
        This method provides a server-side endpoint for consistency.
        
        Returns:
            Success message
        """
        return jsonify({'message': 'Logout successful'})

# Create a Flask blueprint for auth routes
auth_bp = Blueprint('auth', __name__)

# Create controller instance
auth_controller = AuthController()

# Define routes
@auth_bp.route('/login', methods=['POST'])
def login():
    """Route for user login"""
    return auth_controller.login_user()

@auth_bp.route('/logout', methods=['POST'])
def logout():
    """Route for user logout"""
    return auth_controller.logout_user()